### PR TITLE
[5.7] Fixed nested rules in validated data

### PR DESCRIPTION
**Pull request by ttsuru**
Fixed #23708, #24607

In #23708, This is not work with nested array rule.
This PR add 2 types of test cases -> Nested rule (foo.bar) and Nested array rule (foo.*.bar)

### CODE REVIEW

**Comment by jmarcher** [comment]
Can you please add new tests and leave the old tests untouched?

File: src/Illuminate/Foundation/Http/FormRequest.php
Comment (X-Coder264) [inline src/Illuminate/Foundation/Http/FormRequest.php]: This is a major breaking change as I know that this has been used on a lot of projects that I've worked on. Is this even really needed to fix the issue in question?

**Comment by ttsuru** [comment]
@jmarcher 
Already added tests.
I split test into two as existing tests were testing multiple tests in one test.

**Comment by jmarcher** [comment]
@ttsuru What I mean is that you should mostly never change existing tests, this could hide possible breaking changes.

**Comment by X-Coder264** [comment]
Yeah, it's kinda hard to see what's going on with the tests since you changed a bunch of tests from #23708. Please leave those as they were (since they should still pass) and just add a new test to assert that the nested array rules case works as expected.

**Comment by ttsuru** [comment]
@jmarcher Revert tests. But the previous test is not a good.

`array.*` rule is not a commonly used rule.

**Comment by jmarcher** [comment]
Why not? I use it

if you have something like:
<input type="hidden" name="locales[]" value="en">
<input type="hidden" name="locales[]" value="it">
<input type="hidden" name="locales[]" value="de">

You need to validate it as: 

'locales.*' => 'in:en,de,it'

But there are much more uses for it.

**Comment by ttsuru** [comment]
@jmarcher Sorry for not explaining enough.
Only `array.*` rule is not a commonly used.
I think that should be used together `'array' => 'array'`

https://stackoverflow.com/questions/42258185/how-to-validate-array-in-laravel

**Comment by X-Coder264** [comment]
@ttsuru Since you are now calling the `getRules` method on a `\Illuminate\Contracts\Validation\Validator` instance, you should add that method to the interface.

**Comment by ttsuru** [comment]
@X-Coder264 I think so, too.
But even in other places in Laravel, `getRules` and `validate` methods are called.

**Comment by X-Coder264** [comment]
Yeah, that's what happens when people keep adding new features to the framework and they don't add the methods to the interface (even though they should) just so that the MR can be seen as a "no breaking changes" MR which means that instead of having to wait for the feature on the next framework version release, they get the feature "immediately" (just a few days after when the new point release is tagged).

**Comment by ttsuru** [comment]
Can I change `\Illuminate\Contracts\Validation\Validator` interface for 5.7 release ?
Should I add `validate` and  `getRules` method?

**Comment by X-Coder264** [comment]
Yes you can add those methods to the interface, as Laravel allows for breaking changes in "minor" versions since it doesn't follow semver. And if you ask me, yes, they should be added. But the final decision is on Taylor :)

**Comment by ttsuru** [comment]
@X-Coder264 Does Laravel have a policy for changing Contracts?

**Comment by tillkruss** [comment]
@ttsuru: Yes, we only change interfaces in major releases.

**Comment by X-Coder264** [comment]
@tillkruss Yeah, I told him that already 😛

**Comment by ttsuru** [comment]
@tillkruss Thanks. Does Major means 5.6 -> 5.7 or 5.X -> 6.X ?

**Comment by X-Coder264** [comment]
Major means 5.6 -> 5.7. 5.X -> 6.0 is a "paradigm" version according to Laravel terminology.

**Comment by ttsuru** [comment]
Other idea.

`Validator::validate()` returns validated values in #23397
Should I not use `Validator::getRules()`?

**Comment by ttsuru** [comment]
Fixed this issue.

Please review this pull request.

**Review by themsaid — APPROVED** [review]
I think the PR is not breaking the current behaviour, and it fixes the issue reported by the author.

However currently `users.*.name` will return all the content of the "users" input, not just the name, this PR fixes that but people might be relying on the old behaviour so noting this in the docs is needed.

I believe ValidatesRequests::extractInputFromRules() is not needed anymore though.

**Comment by ttsuru** [comment]
Thanks review.
I remove `ValidatesRequests::extractInputFromRules()` method.

**Comment by donnysim** [comment]
I'd say a very common use case is 

```
[
    'foo' => ['required', 'array', 'min:1'],
    'foo.*.name' => ['required'],
    ...
]
```

which will always make the new behavior still not fully usable in those cases as 'foo' will grab all items without filtering fields :(

**Comment by themsaid** [comment]
@donnysim which is kind of expected, the built-in filteration is meant for general usage and it's not meant to cover all edge cases, you'll need to apply special filters to work for some edge cases, that applies to everything in the framework :)

**Comment by donnysim** [comment]
I think it would be better to extract the filtering logic outside of the validator? Then we could use custom extraction on top of the validator extraction? Validator filters 'foo' and then we filter only `['foo.*.name', 'foo.*.last_name']`?

**Comment by themsaid** [comment]
@donnysim you can just ignore it and do your own filtering.

**Pull request merged**
